sound/soc/sof/ipc4-pcm.c

```diff
@@ -191,6 +191,60 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
  * prepare ioctl before the START trigger.
  */
 
+static int sof_ipc4_chain_dma_send_msg(struct snd_sof_dev *sdev,
+				       struct snd_sof_pcm_stream_pipeline_list *pipeline_list,
+				       bool allocate, bool enable)
+{
+	struct sof_ipc4_msg msg = {{ 0 }};
+	int i;
+
+	msg.primary = SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_GLB_CHAIN_DMA);
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
+
+	/*
+	 * Copy the Link DMA ID, host DMA ID, fifo_size and the SCS bit settings from the pipelines.
+	 */
+	for (i = 0; i < pipeline_list->count; i++) {
+		struct snd_sof_pipeline *spipe = pipeline_list->pipelines[i];
+		struct snd_sof_widget *pipe_widget = spipe->pipe_widget;
+		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
```

Comment:
Should we add
```
if (!pipeline->use_chain_dma)
        continue;
```
here?

Reply:
I do not think we should be in this function with a pipeline_list that has any pipeline without pipeline->use_chain_dma. Maybe the check with an error print and return -EINVAL?

Reply:
that check would be redundant and will always be false. No need to add the check

Reply:
Well, I guess we can not check everything all the time.

Reply:
@ranj063 @jsarha But see line 279 to 288
```
	for (i = pipeline_list->count - 1; i >= 0; i--) {
		spipe = pipeline_list->pipelines[i];
		pipe_widget = spipe->pipe_widget;
		pipeline = pipe_widget->private;

		/* If a pipeline in the list uses chained DMA, trigger chained DMA and return */
		if (pipeline->use_chain_dma)
			return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state);
	}
```
We will call `sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state);` if there is one pipeline that use_chain_dma is true.
If I understand it correctly, other pipelines in the pipeline_list could be use_chain_dma = false. Am I correct?
How can we guaranty that all pipelines in the pipeline_list  are all use_chain_dma = true?

Reply:
The quoted loop does not run to the end, so there is really only one check. But yes, I can see now why this is confusing. I'll add a comment here too. In fact even the loop is unnecessary. Checking use_chain_dma from the first element in the pipelist would be enough. However, to make sure the tplg is correct, I keep the loop and then add a check in sof_ipc4_chain_dma_trigger() that all elements in list have the attribute, since that is a requirement for the feature to work.

Reply:
> The quoted loop does not run to the end, so there is really only one check. But yes, I can see now why this is confusing. I'll add a comment here too. In fact even the loop is unnecessary. Checking use_chain_dma from the first element in the pipelist would be enough. However, to make sure the tplg is correct, I keep the loop and then add a check in sof_ipc4_chain_dma_trigger() that all elements in list have the attribute, since that is a requirement for the feature to work.

Thanks @jsarha for the explanation, I understand it now.